### PR TITLE
Net refacto

### PR DESCRIPTION
**Pull request by autra**
Some simplifications of the network layer. I'm wondering if my Fetcher.js should go in another directory (for example `src/Core/Web`) ?

### CODE REVIEW

**Comment by autra** [comment]
@iTowns/reviewers thanks for the review

File: src/Core/Commander/Providers/Fetcher.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            res = resolve;
+            rej = reject;
+        });
+
+        let texture = textureLoader.load(url, res, () => {}, rej);
+        return {texture, promise};
+    }
+}
+
+export default Fetcher;
```
Comment (tbroyer) [inline src/Core/Commander/Providers/Fetcher.js]: Instead of `export default Fetcher` how about exporting each function instead?

``` js
export function json(url) {
   …
}
export function xml(url) {
   …
}
export function texture(url) {
   …
}
```

**Review by tbroyer — COMMENTED** [review]
I approve those changes but will let others "formally" approve them as there might have been reasons to go with all those `IoDriver`s.

File: src/Core/Commander/Providers/Fetcher.js (hunk 1 above)
Comment (peppsac) [reply src/Core/Commander/Providers/Fetcher.js]: Or maybe use a verb instead of a noun : `Fetch.xml` or `Get.xml` ?

**Review by peppsac — CHANGES_REQUESTED** [review]
This PR breaks GPX loading - bisecting indicates that `3ee59564f900617f542162cfd63d962973fb32f3 is the first bad commit`.

To reproduce you can put `agpxfile.gpx` in dist and then: 
`itowns.viewer.loadGPX('http://localhost:8080/dist/agfpxfile.gpx')`

**Comment by peppsac** [comment]
Here's a quick fix:

``` js
-            return new window.DOMParser().parseFromString(response.text(), 'text/xml');
+            return response.text().then(text => new window.DOMParser().parseFromString(text, 'text/xml'));
```

File: src/Core/Commander/Providers/Fetcher.js (hunk 1 above)
Comment (autra) [reply src/Core/Commander/Providers/Fetcher.js]: > how about exporting each function instead?

Oh yeah indeed :-)

> use a verb instead of a noun : Fetch.xml or Get.xml

you mean fetchXml, fetchJson?

Comment (peppsac) [reply src/Core/Commander/Providers/Fetcher.js]: What's the benefit of exporting functions?

I'd prefer `let x = Get.xml(url);` over `let x = xml(url);`.

But you can do the same using `import * as Get from 'Fetcher'`

**Comment by autra** [comment]
@iTowns/reviewers updated. WARNING this PR needs rebasing before merging :-)

**Comment by autra** [comment]
And I forgot to change the export, so definitely not ready to be merged actually :-)

File: src/Core/Commander/Providers/Fetcher.js (hunk 1 above)
Comment (autra) [reply src/Core/Commander/Providers/Fetcher.js]: So yeah, after thinking about this a bit: we must find the right granularity for our exports, neither too big nor too small. Generally speaking, I believe the right level is the *service* as opposed to a single function here. So I would prefer to keep the PR as is.

Then for the naming issue, I like `Fetcher` because we can then refer to it as an entity. Although `Get` is also nice because it gives the code a native language feel, I still have a slight preference for the former, but no strong opinion.

That being said, whether we choose one or the other, I'm not very fond of `import * as Get from 'Fetcher'`: I think it will only be confusing for the reader, who will have to scroll up to check which javascript object we are really talking about.

TL;DR I prefer my way :-D

File: src/Core/Commander/Providers/Fetcher.js [hunk 2]
```diff
@@ -22,7 +22,9 @@ let Fetcher = {
     xml: function(url) {
         return fetch(url).then(response => {
             checkResponse(response);
-            return new window.DOMParser().parseFromString(response.text(), 'text/xml');
+            return response.text();
+        }).then(text => {
+            return new window.DOMParser().parseFromString(text, 'text/xml');
```
Comment (tbroyer) [inline src/Core/Commander/Providers/Fetcher.js]: nit: you could remove the braces and the `return` here.

File: src/Core/Commander/Providers/Fetcher.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    if (response.status < 200 || response.status >= 300) {
+        var error = new Error(`Error loading ${response.url}: status ${response.status}`);
+        error.status = response.status;
+        throw error;
+    }
+}
+
+let Fetcher = {
+
+    json: function(url) {
```
Comment (tbroyer) [inline src/Core/Commander/Providers/Fetcher.js]: nit: how about the "method definition" syntax?

```js
{
  json(url) { … }
  xml(url) { … }
  texture(url) { … }
}
```

File: src/Core/Commander/Providers/Fetcher.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+function checkResponse(response) {
+    if (response.status < 200 || response.status >= 300) {
+        var error = new Error(`Error loading ${response.url}: status ${response.status}`);
+        error.status = response.status;
+        throw error;
+    }
+}
+
+let Fetcher = {
```
Comment (tbroyer) [inline src/Core/Commander/Providers/Fetcher.js]: nit: does the object need to be named?

```js
export default {
  …
};
```

File: src/Core/Commander/Providers/Fetcher.js (hunk 1 above)
Comment (tbroyer) [reply src/Core/Commander/Providers/Fetcher.js]: Wrt. naming, `Fetch` vs. `Fetcher`, can't decide…

**Review by tbroyer — APPROVED** [review]
SGTM, only suggesting stylistic tweaks.

(BTW, sorry for the spam, I keep clicking "Add single comment" instead of "start review"; don't know why…)

File: src/Core/Commander/Providers/Fetcher.js (hunk 3 above)
Comment (autra) [reply src/Core/Commander/Providers/Fetcher.js]: Yep, and actually I need to do it this way for the linter in #212 :-)

**Comment by autra** [comment]
@tbroyer @peppsac updated. Don't forget, I need to squash before merging :-)

File: src/Core/Commander/Providers/Fetcher.js [hunk 5]
```diff
@@ -39,5 +39,3 @@ export default {
         return {texture, promise};
     }
 }
-
-export default Fetcher;
```
Comment (tbroyer) [inline src/Core/Commander/Providers/Fetcher.js]: I believe that changed line should go in the previous commit.

Comment (autra) [reply src/Core/Commander/Providers/Fetcher.js]: Yep, updated.

**Review by peppsac — APPROVED** [review]
LGTM - except I still prefer the Get.xml() form but it's not a big deal :)

**Comment by autra** [comment]
I have squashed the fixup. Is this ready to be merged?

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 6]
```diff
@@ -134,7 +133,7 @@ ApiGlobe.prototype.addImageryLayer = function(layer) {
  */
 
 ApiGlobe.prototype.addImageryLayerFromJSON = function(url) {
-     return JSONDriver.read(url).then(function(result){
+     return Fetcher.json(url).then(function(result){
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: add space before {

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 7]
```diff
@@ -238,7 +237,7 @@ ApiGlobe.prototype.addElevationLayer = function(layer) {
  */
 
 ApiGlobe.prototype.addElevationLayersFromJSON = function(url) {
-     return JSONDriver.read(url).then(function(result){
+     return Fetcher.json(url).then(function(result){
```
Comment (gchoqueux) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: add space before {

File: src/Core/Commander/Providers/Fetcher.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        let res, rej;
+        let promise = new Promise(function(resolve, reject) {
+            res = resolve;
+            rej = reject;
+        });
+
+        let texture = textureLoader.load(url, res, () => {}, rej);
+        return {texture, promise};
+    }
+}
```
Comment (gchoqueux) [inline src/Core/Commander/Providers/Fetcher.js]: add semicolon

File: src/Core/Commander/Providers/WMS_Provider.js [hunk 9]
```diff
@@ -109,32 +106,21 @@ WMS_Provider.prototype.getColorTexture = function(tile, layer, bbox, pitch) {
...
-            result.texture = texture;
-        else
-        {
-            result.texture = new THREE.Texture(image);
-            result.texture.needsUpdate = true;
-            result.texture.generateMipmaps = false;
-            result.texture.magFilter = THREE.LinearFilter;
-            result.texture.minFilter = THREE.LinearFilter;
-            result.texture.anisotropy = 16;
-            result.texture.url = url;
```
Comment (gchoqueux) [inline src/Core/Commander/Providers/WMS_Provider.js:122]: Why you remove `result.texture.url = url;`?

File: src/Core/Commander/Providers/WMTS_Provider.js [hunk 10]
```diff
@@ -200,39 +197,26 @@ WMTS_Provider.prototype.getColorTexture = function(coWMTS, pitch, layer) {
...
-        if (texture)
-            result.texture = texture;
-        else {
-            result.texture = new THREE.Texture(image);
-            result.texture.needsUpdate = true;
-            result.texture.generateMipmaps = false;
-            result.texture.magFilter = THREE.LinearFilter;
-            result.texture.minFilter = THREE.LinearFilter;
-            result.texture.anisotropy = 16;
-            result.texture.url = url;
```
Comment (gchoqueux) [inline src/Core/Commander/Providers/WMTS_Provider.js:216]: Why you remove `result.texture.url = url;`?

File: src/Core/Commander/Providers/Fetcher.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
+import { TextureLoader } from 'three';
```
Comment (gchoqueux) [inline src/Core/Commander/Providers/Fetcher.js:1]: I know that THREE is present everywhere in code, but it'll be a good idea to limit its expansion in iTowns.

**Review by gchoqueux — CHANGES_REQUESTED** [review]

**Comment by gchoqueux** [comment]
There's a bug in globe loading : some tiles flick (texture->blue->texture).

File: src/Core/Commander/Providers/Fetcher.js (hunk 11 above)
Comment (autra) [reply src/Core/Commander/Providers/Fetcher.js:1]: What worries you exactly?

Comment (gchoqueux) [reply src/Core/Commander/Providers/Fetcher.js:1]: This is not yet the case for the moment, but a good principle will be not to contaminate our code with that of THREE.js. (Especially for loading)
The Fetcher can return an image and not a texture.

**Comment by autra** [comment]
> There's a bug in globe loading : some tiles flick (texture->blue->texture).

Yep indeed. I'm looking at it.

File: src/Core/Commander/Providers/WMS_Provider.js (hunk 9 above)
Comment (autra) [reply src/Core/Commander/Providers/WMS_Provider.js:122]: I thought it was not used anywhere, but I'll double-check

Comment (autra) [reply src/Core/Commander/Providers/WMS_Provider.js:122]: So yeah, I couldn't see any use of this parameter in the code, but I may have missed something. Don't hesitate to point potential problems to me.

File: src/Core/Commander/Providers/Fetcher.js (hunk 11 above)
Comment (autra) [reply src/Core/Commander/Providers/Fetcher.js:1]: > (Especially for loading)

What do you mean by that? Importing it one or n times won't change the loading time, it's just a scope thing AFAIK (hopefully so, otherwise `import` syntax would have to be avoided at all cost!)

>  The Fetcher can return an image and not a texture.

That would mean nearly exactly duplicating what Three.js already does (see [TextureLoader](https://github.com/mrdoob/three.js/blob/6c7f000734f8579da37fb39e5c2e9e5e2dfb14f8/src/loaders/TextureLoader.js)). So a bigger code (so slower to load), that we need to maintain.

I feel like I'm missing the point: why importing and using an utility function is "contaminating the code"? And why do we want to limit the "expansion of three.js in itowns"? (I don't know what you mean by "expansion" and "contamination" btw). Do we plan on being agnostic regarding the WebGL lib?

Or do you mean you feel this function should be elsewhere?

Comment (peppsac) [reply src/Core/Commander/Providers/Fetcher.js:1]: I'm with @autra on this one.
I prefer to take full advantage of our chosen framework (three.js). We already have way bigger dependencies (Mesh, Materials, Vectors) so I don't see the value of *not* using the `TextureLoader` function.
Actually I'd rather drop our limited reimplementation of three.js features: `Camera` and `Scene` for instance but that's another story :)

**Comment by autra** [comment]
@gchoqueux I corrected the tile flicking and the nits.

File: src/Core/Commander/Providers/Fetcher.js (hunk 11 above)
Comment (gchoqueux) [reply src/Core/Commander/Providers/Fetcher.js:1]: > We already have way bigger dependencies (Mesh, Materials, Vectors)

It's true and normal, the use of THREE should only encapsulate in objects in Renderer folder.
I would like to have the minimum use of THREE in folders Scene, Core and Globe. It would be better in terms of design.

Ok for the moment, it's not very important, so keep textureLoader in Fetcher

File: src/Core/Commander/Providers/WMS_Provider.js (hunk 9 above)
Comment (gchoqueux) [reply src/Core/Commander/Providers/WMS_Provider.js:122]: Ok, remove `url`

File: src/Core/Commander/Providers/WMTS_Provider.js (hunk 10 above)
Comment (gchoqueux) [reply src/Core/Commander/Providers/WMTS_Provider.js:216]: Ok, remove `url`

**Review by gchoqueux — APPROVED** [review]

**Comment by gchoqueux** [comment]
r+
